test/testscript.go

```diff
@@ -32,27 +32,70 @@ func NewProjectTest(t *testing.T, projectDir string) *ProjectTest {
 	}
 }
 
+func copyFile(srcPath string, dstPath string, mode os.FileMode) error {
+	srcContent, err := os.ReadFile(srcPath)
+	if err != nil {
+		return fmt.Errorf("reading source file %s: %w", srcPath, err)
+	}
+
+	if err := os.WriteFile(dstPath, srcContent, mode); err != nil {
+		return fmt.Errorf("writing destination file %s: %w", dstPath, err)
+	}
+	return nil
+}
+
+func makeDir(dirPath string, mode os.FileMode) error {
+	if err := os.MkdirAll(dirPath, mode); err != nil {
+		return fmt.Errorf("creating directory %s: %w", dirPath, err)
+	}
+	return nil
+}
+
 func (e *ProjectTest) CopyFile(srcRelPath string) *ProjectTest {
 	setupFunc := func(env *testscript.Env) error {
 		src := filepath.Join(e.projectDir, srcRelPath)
 		dst := filepath.Join(env.WorkDir, srcRelPath)
 		dstDir := filepath.Dir(dst)
-		if err := os.MkdirAll(dstDir, 0755); err != nil {
-			msg := "failed to create destination directory %s: %w"
-			return fmt.Errorf(msg, dstDir, err)
+
+		if err := makeDir(dstDir, 0755); err != nil {
+			return err
 		}
+		return copyFile(src, dst, 0644)
```

Comment:
What if we want to copy an executable file?

Reply:
I had the same question when writing this but kicked it down the road. We could create another version of the copy fn that allows specification of the file permissions. Either that or change the testscript code to simply keep the existing file permissions which could be a simpler behavior.

Reply:
FWIW, I would suggest keeping the permissions as they were provided is a good default option.

Reply:
Updated to preserve file and directory permissions when copying